README.md
# react-variable-height-infinite-scroller

An infinite scroller especially made for variable row heights (no precomputation of row height necessary).

See a [Demo](http://tnrich.github.io/react-variable-height-infinite-scroller/)

# Why

Because sometimes you don't know the size of the rows you're going to render before rendering

# Install: 

```sh
npm i --save react-variable-height-infinite-scroller
```

# Usage:

```
<InfiniteScroller
  averageElementHeight={100} // this is a guess you make!
  containerHeight={600}
  rowToJumpTo={rowToJumpTo} // (optional) set this if you want to start/jump to a a particular row. Must be passed as a new object each time to allow for difference checking
  renderRow={renderRow} // function to render a row
  totalNumberOfRows={fakeRows.length} // an array of data for your rows
/>
```

## Props

| Name | Default | Description |
| :--- |:---:| :---|
| containerHeight `number` | __required__ | Maximum height of the scroll container |
| totalNumberOfRows `number` | __required__ | Length of the data array |
| renderRow `function` | __required__ | Function to render a row |
| averageElementHeight `number` | *(optional)* 100 | This is a guess you make of what the average row height will be height. This is used to approximate how far to move the scrollbar |
| rowToJumpTo | *(optional)* 0 | Object of shape `{ row: Number }`. Row you want to start with or jump to. Must be passed as a new object each time to allow for difference checking |
| jumpToBottomOfRow `boolean` | *(optional)* false |  By default jumping to a row jumps to the top. Set to true if you want to jump to the bottom of the row |
| containerClassName `string` | *(optional)* infiniteContainer | className to apply on container |
| onScroll `function` | *(optional)* no-op |  Hook to call on scroll |

## Methods
| Name | Description |
| :--- |:---|
| scrollTo | Alternate way to jump to a particular row. Usage below. |
### scrollTo usage*:

```js
class MyComponent extends Component {
  scrollSomewhere() {
    this.list.scrollTo(Math.floor(Math.random() * 100));
  }

  render() {
    return (
      <div>
        Here is my list:
        <InfiniteScroller length={100} itemRenderer={...} ref={c => this.list = c} />
        <button onClick={::this.scrollSomewhere}>Scroll Somewhere</button>
      </div>
    );
  }
}
```

Taken from the demo code: 

```javascript
import React from 'react';
import InfiniteScroller from './InfiniteScroller.js';

function getFakeRowsWithHeights(numberOfRows) {
  let newFakeRows = [];
  for (let i = 0; i < numberOfRows; i++) {
    newFakeRows.push({height: Math.floor(1000 * Math.random())});
  }
  return newFakeRows;
}

const Example1 = React.createClass({
  getNewRandomRow(totalRows) {
    return {row: Math.floor(totalRows * Math.random())};
  },

  getInitialState() {
    const newNumberOfRowsToDisplay = Math.floor(Math.random() * 200);
    const newFakeRows = getFakeRowsWithHeights(newNumberOfRowsToDisplay);
    return {
      rowToJumpTo: null,
      newRowToJumpTo: this.getNewRandomRow(newFakeRows.length),
      fakeRows: newFakeRows,
    };
  },
  render() {
    const newNumberOfRowsToDisplay = Math.floor(Math.random() * 200);
    return (
      <div style={{width: 300}} overflow="scroll">
        <h3>
        Example 1: Random number of rows and row heights
        </h3>
        <button onClick={() => {
          this.setState({
            rowToJumpTo: this.state.newRowToJumpTo,
            newRowToJumpTo: this.getNewRandomRow(this.state.fakeRows.length),
          });
        }}>
          Jump to a random row: Row #{this.state.newRowToJumpTo.row} (its height is {this.state.fakeRows[this.state.newRowToJumpTo.row].height})
        </button>
        <button onClick={() => {
          const newFakeRows = getFakeRowsWithHeights(newNumberOfRowsToDisplay);
          this.setState({
            fakeRows: newFakeRows,
            newRowToJumpTo: this.getNewRandomRow(newFakeRows.length),
          });
        }}>
          Create {newNumberOfRowsToDisplay} new rows
        </button>

        <InfiniteScroller
          averageElementHeight={100} // this is a guess you make!
          containerHeight={600}
          rowToJumpTo={this.state.rowToJumpTo} // (optional) set this if you want to start/jump to a a particular row. Must be passed as a new object each time to allow for difference checking
          renderRow={this.renderRow} // function to render a row
          totalNumberOfRows={this.state.fakeRows.length} // an array of data for your rows
        />
      </div>
    );
  },

  renderRow(rowNumber) {
    const heightOfRow = this.state.fakeRows[rowNumber].height;
    return (
      <div
        key={rowNumber}
        style={{height: heightOfRow, background: rowNumber % 2 === 0 ? 'red' : 'orange'}}
      >
        Height: {heightOfRow}
        <br/>
        Row Number: {rowNumber}
      </div>
    );
  },
});

React.render(<Example1 />, document.getElementById('container'));
```

# Contributing

[Changelog](https://github.com/tnrich/react-variable-height-infinite-scroller/blob/master/CHANGELOG.md) is now autogenerated. So commits have to be prefixed by one the four following prefixes to get added to the changelog: 

- `[added]` added a new feature
- `[changed]` changed an existing feature
- `[fixed]` fixed a bug
- `[removed]` removed something or a file

Run ```npm test``` to lint

#Changelog: 
[Changelog](https://github.com/tnrich/react-variable-height-infinite-scroller/blob/master/CHANGELOG.md)

*(taken from "react-list")


CHANGELOG.md
v3.3.5 - Wed, 01 Jun 2016 00:59:44 GMT
--------------------------------------

- [99359dc](../../commit/99359dc) [added] scrollTo method
- [8bfbd07](../../commit/8bfbd07) [removed] react and react-dom as peer dependencies



v3.3.3 - Wed, 13 Jan 2016 07:33:19 GMT
--------------------------------------

- [0f4e0cf](../../commit/0f4e0cf) [fixed] error with jumping to a row when dealing with a  small number of rows + small row heights relative to the size of container



v3.3.1 - Sat, 02 Jan 2016 08:27:05 GMT
--------------------------------------

- [13acf96](../../commit/13acf96) [fixed] loosened peer dependency ranges



v3.3.0 - Sat, 02 Jan 2016 08:15:40 GMT
--------------------------------------





v3.2.0 - Sat, 02 Jan 2016 08:15:00 GMT
--------------------------------------

- [8be1599](../../commit/8be1599) [fixed] findDomNode call, upgraded deps



v3.1.1 - Tue, 27 Oct 2015 17:56:13 GMT
--------------------------------------





v3.1.0 - Tue, 27 Oct 2015 17:39:32 GMT
--------------------------------------

- [04764f8](../../commit/04764f8) [added] ability to jump to the bottom of a row instead of the top



v3.0.1 - Tue, 27 Oct 2015 04:40:28 GMT
--------------------------------------





v3.0.0 - Tue, 27 Oct 2015 04:36:33 GMT
--------------------------------------

- [dbb41ec](../../commit/dbb41ec) [fixed] linting errors
- [7b35e5d](../../commit/7b35e5d) [fixed] scroll jump to bottom row



v2.0.9 - Fri, 02 Oct 2015 18:21:59 GMT
--------------------------------------





v2.0.7 - Fri, 04 Sep 2015 18:04:00 GMT
--------------------------------------





v2.0.6 - Fri, 04 Sep 2015 08:15:14 GMT
--------------------------------------

- [c6a760f](../../commit/c6a760f) [fixed] leftover console.log



v2.0.5 - Fri, 04 Sep 2015 08:06:05 GMT
--------------------------------------

- [b39d143](../../commit/b39d143) [fixed] row start adjustments now calculated based on estimated row heights instead of average



v2.0.4 - Tue, 01 Sep 2015 19:26:42 GMT
--------------------------------------

- [ab165fa](../../commit/ab165fa) [fixed] default props has to be a function
- [c0dd3b8](../../commit/c0dd3b8) [fixed] correcting error in previous commit
- [4c67f39](../../commit/4c67f39) [fixed] bug with example jumping to line code
- [00f12c9](../../commit/00f12c9) [added] propType table in the README
- [161ee0b](../../commit/161ee0b) [added] containerClassName, allows custom className on container
- [b20f145](../../commit/b20f145) [added] onScroll callback prop



v2.0.3 - Mon, 31 Aug 2015 18:42:25 GMT
--------------------------------------

- [2158398](../../commit/2158398) [added] automated release tools
- [90cc6fb](../../commit/90cc6fb) [added] eslint and changed to use babel




example.js
import React from 'react';
import ReactDOM from 'react-dom';

import InfiniteScroller from './InfiniteScroller.js';
import Example2 from './example2.js';
import Example3 from './example3.js';
import Example4 from './example4.js';

function getFakeRowsWithHeights(numberOfRows) {
  let newFakeRows = [];
  for (let i = 0; i < numberOfRows; i++) {
    newFakeRows.push({height: Math.floor(1000 * Math.random())});
  }
  return newFakeRows;
}

const Example1 = React.createClass({
  getNewRandomRow(totalRows) {
    return {row: Math.floor(totalRows * Math.random())};
  },

  getInitialState() {
    const newNumberOfRowsToDisplay = Math.floor(Math.random() * 200);
    const newFakeRows = getFakeRowsWithHeights(newNumberOfRowsToDisplay);
    return {
      rowToJumpTo: null,
      newRowToJumpTo: this.getNewRandomRow(newFakeRows.length),
      fakeRows: newFakeRows,
    };
  },
  render() {
    const newNumberOfRowsToDisplay = Math.floor(Math.random() * 200);
    return (
      <div style={{width: 300}} overflow="scroll">
        <h3>
        Example 1: Random number of rows and row heights
        </h3>
        <button onClick={() => {
          this.setState({
            rowToJumpTo: this.state.newRowToJumpTo,
            newRowToJumpTo: this.getNewRandomRow(this.state.fakeRows.length),
          });
        }}>
          Jump to a random row: Row #{this.state.newRowToJumpTo.row} (its height is {this.state.fakeRows[this.state.newRowToJumpTo.row].height})
        </button>
        <button onClick={() => {
          this.list.scrollTo(this.state.newRowToJumpTo.row);
          this.setState({
            newRowToJumpTo: this.getNewRandomRow(this.state.fakeRows.length),
          });
        }}>
          Jump to a random row using the scrollTo interface: Row #{this.state.newRowToJumpTo.row} (its height is {this.state.fakeRows[this.state.newRowToJumpTo.row].height})
        </button>
        <button onClick={() => {
          const newFakeRows = getFakeRowsWithHeights(newNumberOfRowsToDisplay);
          this.setState({
            fakeRows: newFakeRows,
            newRowToJumpTo: this.getNewRandomRow(newFakeRows.length),
          });
        }}>
          Create {newNumberOfRowsToDisplay} new rows
        </button>

        <InfiniteScroller
          averageElementHeight={100} // this is a guess you make!
          ref={c => this.list = c}
          containerHeight={600}
          rowToJumpTo={this.state.rowToJumpTo} // (optional) set this if you want to start/jump to a a particular row. Must be passed as a new object each time to allow for difference checking
          renderRow={this.renderRow} // function to render a row
          totalNumberOfRows={this.state.fakeRows.length} // an array of data for your rows
        />
      </div>
    );
  },

  renderRow(rowNumber) {
    const heightOfRow = this.state.fakeRows[rowNumber].height;
    return (
      <div
        key={rowNumber}
        style={{height: heightOfRow, background: rowNumber % 2 === 0 ? 'red' : 'orange'}}
      >
        Height: {heightOfRow}
        <br/>
        Row Number: {rowNumber}
      </div>
    );
  },
});

ReactDOM.render((
  <div style={{display: 'flex'}} >
    <Example1 />
    <div style={{width: 100}}/>
    <Example2 />
    <div style={{width: 100}}/>
    <Example3 />
    <div style={{width: 100}}/>
    <Example4 />
  </div>
  ), document.getElementById('container'));
